### PR TITLE
Retry hostname->IP: [Errno -2] Name or service not known

### PR DESCRIPTION
**Pull request by ptabor**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

During cluster configuration, the hostname is getting resolved to IP,
as etcd requires IP address as listening address.

1. Due to connectivity flakes or delayed network inititalization, sometimes the IP fails to be resolved to a name with following error:
```
[Errno -2] Name or service not known
```
that leads to attempt to run etcd with empty flag.

The PR adds a proper retry (up to 5 minutes) in case the connectivity problems happens.

I considered alternatives like: `getent hosts foo`, but unfortunetelly thay can return IPv6 that etcd is not ready for (yet).

#### Which issue(s) this PR fixes:

NONE

#### Does this PR introduce a user-facing change?

NONE

### CODE REVIEW

**Comment by mm4tt** [comment]
Is this still needed with changes we added in https://github.com/kubernetes/kubernetes/pull/101176? In particular, we changed the `kube-master-installation.service` to include `Requires=network-online.target`. Wouldn't it guarantee that network is up when the `configure-helper.sh` is running?

**Comment by mm4tt** [comment]
@jkh52, could you shed more light on this? I think you've spent a lot of time debugging similar issues

**Comment by ptabor** [comment]
`Requires=network-online.target` should definitely help. 

But this is still dependency on network connectivity, availability of DNS server, etc. that might flake and for in-depth reliability such calls should have a retry policy.

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -1781,7 +1781,27 @@ function start-kube-proxy {
 # $5: pod name, which should be either etcd or etcd-events
 function prepare-etcd-manifest {
   local host_name=${ETCD_HOSTNAME:-$(hostname -s)}
-  local -r host_ip=$(python3 -c "import socket;print(socket.gethostbyname(\"${host_name}\"))")
+
```
Comment (wojtek-t) [inline cluster/gce/gci/configure-helper.sh:1784]: Why not doing what we did internally instead (i.e. something like:

lo

File: cluster/gce/gci/configure-helper.sh [hunk 2]
```diff
@@ -1781,7 +1781,27 @@ function start-kube-proxy {
...
+      print(socket.gethostbyname(host))
+      break
+    except Exception as e:
+      sys.stderr.write("error: resolving host %s to IP failed: %s\n" % (host, e))
+      time.sleep(1)
+      continue
+EOM
+
+  local nl=$'\n'
+  local -r host_ip=$(python3 -c "${resolve_host_script_py};${nl}${nl}resolve(\"${host_name}\")")
```
Comment (wojtek-t) [inline cluster/gce/gci/configure-helper.sh:1804]: Why not doing what we did internally instead (i.e. something like:

host_ip=$(ETCD_IP:-127.0.0.1)

?
[and allow setting ETCD_IP explicitly if needed]

Comment (ptabor) [reply cluster/gce/gci/configure-helper.sh:1804]: I assume it would be a breaking change to all users that do set ETCD_HOSTNAME explicitly already
and depend on auto-resolution of this variable to IP address. 

But I don't know all the usage modes, so tried to make as safe change as possible, that can be even backported. 

If we want to go that path, I would consider: 
```
host_ip=$(ETCD_IP:hostname -i)
```
for symmetry with `hostname -s`

**Comment by ptabor** [comment]
```
{"step_name":"StartEtcdServers","status":"STARTED","status_reason":"","latency":null}}
May 19 12:32:01.139759 e2e-3322bca04a-a7d53-master configure-helper.sh[9594]: Start etcd pods
May 19 12:32:01.478776 e2e-3322bca04a-a7d53-master configure-helper.sh[9594]: [cloud.kubernetes.monitoring.proto.SerialportLog] {"cluster_hash":"get-metadata-value-error","vm_instance_name":"e2e-3322bca04a-a7d53-master","boot_id":"1814ebbe25ca4ead877c64a504614ade","timestamp":"2021-05-19T12:32:01.469917707Z","bootstrap_status":{"step_name":"StartEtcdServers","status":"COMPLETED","status_reason":"","latency":"0.334141s"}}
``` passed in the recent version.

**Comment by ptabor** [comment]
The test fails with: 

```
May 19 19:32:22.974545 e2e-d984be3f96-a7d53-master configure-helper.sh[11647]: error: the server doesn't have a resource type "volumesnapshotclasses"
May 19 19:32:33.290563 e2e-d984be3f96-a7d53-master configure-helper.sh[11769]: error: the server doesn't have a resource type "volumesnapshotclasses"
May 19 19:32:43.547266 e2e-d984be3f96-a7d53-master configure-helper.sh[11910]: error: the server doesn't have a resource type "volumesnapshotclasses"
May 19 19:32:54.086932 e2e-d984be3f96-a7d53-master configure-helper.sh[12020]: No resources found
May 19 19:32:54.509486 e2e-d984be3f96-a7d53-master configure-helper.sh[12025]: No resources found
May 19 19:32:55.124737 e2e-d984be3f96-a7d53-master configure-helper.sh[12029]: No resources found in default namespace.
May 19 19:32:55.126372 e2e-d984be3f96-a7d53-master configure-helper.sh[11221]: Wait until volume snapshot RBAC rules are installed
May 19 19:32:55.484545 e2e-d984be3f96-a7d53-master configure-helper.sh[12053]: NAME                              ROLE                                            AGE
May 19 19:32:55.485098 e2e-d984be3f96-a7d53-master configure-helper.sh[12053]: volume-snapshot-controller-role   ClusterRole/volume-snapshot-controller-runner   5s
May 19 19:32:55.486454 e2e-d984be3f96-a7d53-master configure-helper.sh[11221]: Wait until volume snapshot controller is installed
May 19 19:32:55.881489 e2e-d984be3f96-a7d53-master configure-helper.sh[12057]: Error from server (NotFound): statefulsets.apps "volume-snapshot-controller" not found
May 19 19:33:06.034672 e2e-d984be3f96-a7d53-master configure-helper.sh[12194]: Error from server (NotFound): statefulsets.apps "volume-snapshot-controller" not found
May 19 19:33:16.314812 e2e-d984be3f96-a7d53-master configure-helper.sh[12332]: Error from server (NotFound): statefulsets.apps "volume-snapshot-controller" not found
```

But it doesn't seem to be related to my PR

**Comment by ptabor** [comment]
Note: I changed also IP to 127.0.0.1 for single-master clusters in order to limit port-exposure on the host level.

File: cluster/gce/manifests/etcd.manifest [hunk 3]
```diff
@@ -27,7 +27,7 @@
...
+              "set -o errexit; set -x; declare -p; if [ -e /usr/local/bin/migrate-if-needed.sh ]; then /usr/local/bin/migrate-if-needed.sh 1>>/var/log/etcd{{ suffix }}.log 2>&1; fi; exec /usr/local/bin/etcd --name etcd-{{ hostname }} --listen-peer-urls {{ etcd_protocol }}://{{ host_ip }}:{{ server_port }} --initial-advertise-peer-urls {{ etcd_protocol }}://{{ hostname }}:{{ server_port }} --advertise-client-urls {{ etcd_apiserver_protocol }}://127.0.0.1:{{ port }} --listen-client-urls {{ etcd_apiserver_protocol }}://{{ listen_client_ip }}:{{ port }} {{ quota_bytes }} --data-dir /var/etcd/data{{ suffix }} --initial-cluster-state {{ cluster_state }} --initial-cluster {{ etcd_cluster }} {{ etcd_creds }} {{ etcd_apiserver_creds }} {{ etcd_extra_args }} 1>>/var/log/etcd{{ suffix }}.log 2>&1"
```
Comment (mm4tt) [inline cluster/gce/manifests/etcd.manifest:30]: I'd revert changes in this file. They are no-op without adding something like:
```
exec 1>>/var/log/etcd{{ suffix }}.log
```

See example etcd log from on of presubmits - 
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/101781/pull-kubernetes-e2e-gce-ubuntu-containerd/1395653211989217280/artifacts/e2e-5c6e78df97-a7d53-master/etcd.log

If we want to change it let's do it in a separate PR

File: cluster/gce/gci/configure-helper.sh [hunk 4]
```diff
@@ -1771,7 +1771,33 @@ function start-kube-proxy {
...
+      time.sleep(1)
+      continue
+
+'
+  if [[ -n "${MULTIMASTER:-}" ]]; then
+    # Multi-master should expose the peer server on the host level.
+    local -r host_ip=$(python3 -c "${resolve_host_script_py}"$'\n'"resolve(\"${host_name}\")")
+  else
+    # Single-master should expose the peer connection on the host level.
+    local host_ip=127.0.0.1
```
Comment (mm4tt) [inline cluster/gce/gci/configure-helper.sh:1798]: Forgive my bash ignorance, but why we don't declare this one with -r?

Comment (ptabor) [reply cluster/gce/gci/configure-helper.sh]: > -r	to make NAMEs readonly

We don't have consistent style of using it, but that's more 'defensive'

File: cluster/gce/manifests/etcd.manifest (hunk 3 above)
Comment (ptabor) [reply cluster/gce/manifests/etcd.manifest]: Done

File: cluster/gce/gci/configure-helper.sh (hunk 4 above)
Comment (ptabor) [reply cluster/gce/gci/configure-helper.sh]: Done

**Comment by jkh52** [comment]
> @jkh52, could you shed more light on this? I think you've spent a lot of time debugging similar issues
Sorry for late reply; @avrittrohwer took over my debugging.

Agree that the systemd improvement seems relevant.
Also see the internal issue at b/189264693.

**Comment by mm4tt** [comment]
@ptab, sorry for the delay. It went off my radar.

Have we tested the first commit (introducing the new python script) without the second one? Unfortunately, our CI tests don't set up MULTIMASTER clusters. If we're submitting these two commits together, and there is a bug in the new python script, the presubmit won't catch it. 

If yes, then I'm fine with merging this.

If not, can we move the second commit to a separate PR and submit it once this one merges?

**Comment by mm4tt** [comment]
@ptabor ^ (I originally used a wrong gtihub handle, not sure if you got a notification)

**Comment by ptabor** [comment]
@mm4tt 
I tested it locally. Also tested in isolation: 
```
#!/bin/bash

function prepare-etcd-manifest {
  local host_name=${ETCD_HOSTNAME:-$(hostname -s)}

  local resolve_host_script_py='
import socket
import time
import sys
timeout_sec=300
def resolve(host):
  for attempt in range(timeout_sec):
    try:
      print(socket.gethostbyname(host))
      break
    except Exception as e:
      sys.stderr.write("error: resolving host %s to IP failed: %s\n" % (host, e))
      time.sleep(1)
      continue
'

  local -r host_ip=$(python3 -c "${resolve_host_script_py}"$'\n'"resolve(\"${host_name}\")")
  echo "$host_ip"
}

prepare-etcd-manifest
```

Still isolated the multi-cluster change to a separate PR.

**Comment by aojea** [comment]
> I considered alternatives like: `getent hosts foo`, but unfortunetelly thay can return IPv6 that etcd is not ready for (yet).

:thinking: etcd does support IPv6 , doesn't it? there are IPv6 jobs running it in ipv6 in KIND

If you still want IPv4, have you tried `getent ahostsv4 hostname`?

**Comment by ptabor** [comment]
> > I considered alternatives like: `getent hosts foo`, but unfortunetelly thay can return IPv6 that etcd is not ready for (yet).
> 
>  etcd does support IPv6 , doesn't it? there are IPv6 jobs running it in ipv6 in KIND

@aojea
TBH: etcd does not have any tests of IPv6. I'm afraid of running it without explicit coverage including mtls hostname resolution.  
 
> If you still want IPv4, have you tried `getent ahostsv4 hostname`?

I thought about this but it returns multiple addresses. Whole solution would have similar complexity (retry loop, grep), but additional risk coming from the potential change of the behavior.

**Comment by aojea** [comment]
> @aojea
> TBH: etcd does not have any tests of IPv6. I'm afraid of running it without explicit coverage including mtls hostname resolution.

you know better than me for sure :smile: , if you give me more details about the code/area that you want to improve and some guidance I'm happy to work on that ...

**Comment by pacoxu** [comment]
> @ptabor: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

/release-note-none
